mlflow/langchain/api_request_parallel_processor.py

```diff
@@ -351,6 +371,63 @@ def _try_transform_response_to_chat_format(response):
         else:
             return transformed_response.model_dump(mode="json")
 
+    @staticmethod
+    def _try_transform_response_iter_to_chat_format(chunk_iter):
+
+        def _gen_converted_chunk(message_content, finish_reason):
+            transformed_response = _ChatChunkResponse(
+                id=None,
+                created=int(time.time()),
+                model=None,
+                choices=[
+                    _ChatChoiceDelta(
+                        index=0,
+                        delta=_ChatDeltaMessage(
+                            role="assistant",
+                            content=message_content,
+                        ),
+                        finish_reason=finish_reason,
+                    )
+                ],
+            )
+
+            if Version(pydantic.__version__) < Version("2.0"):
+                return json.loads(transformed_response.json())
+            else:
+                return transformed_response.model_dump(mode="json")
+
+        def _convert(chunk):
+            if isinstance(chunk, str):
+                message_content = chunk
+                finish_reason = None
+                need_append_finish_chunk = True
+            elif isinstance(chunk, AIMessageChunk):
+                message_content = chunk.content
+                finish_reason = chunk.response_metadata.get('finish_reason')
+                need_append_finish_chunk = False
+            elif isinstance(chunk, AIMessage):
+                # The langchain chat model does not support stream
+                # so `model.stream` returns the whole result.
+                message_content = chunk.content
+                finish_reason = 'stop'
+                need_append_finish_chunk = False
+            else:
+                return chunk, False
+            return (
+                _gen_converted_chunk(message_content, finish_reason=finish_reason),
+                need_append_finish_chunk,
+            )
+
+        def _result_gen_fn():
+            need_append_finish_chunk = False
```

Comment:
What is `need_append_finish_chunk`?

Reply:
Because we need to generate an iterator of `_ChatChunkResponse`, and the last `_ChatChunkResponse` object must have `finish_reason` = `stop`,

but if the input iterator is an iterator of string, we can't easily generate the last chunk with `finish_reason` = `stop`, because we can't judge whether it is the last chunk in the iterator.

As a workaround, in this case, I append a last empty chunk with `finish_reason` = `stop`,
See:

```
            if need_append_finish_chunk:
                yield _gen_converted_chunk('', finish_reason="stop")
```

Reply:
> the last _ChatChunkResponse object must have finish_reason = stop,

Where does this constraint come from?

Reply:
We can see https://platform.openai.com/docs/api-reference/making-requests

> Now that you've generated your first chat completion, let's break down the [response object](https://platform.openai.com/docs/api-reference/chat/object). We can see the finish_reason is stop which means the API returned the full chat completion generated by the model without running into any limits. In the choices list, we only generated a single message but you can set the n parameter to generate multiple messages choices.

'stop' means it completes without other error.

Reply:
what would happen if a stream ends without finish reason?

Reply:
> what would happen if a stream ends without finish reason?

I don't think this is a valid case, if a stream ends, it should have a finish reason 'stop' (completed) or other finish reason e.g. 'length' (generation exceeded max token limit)

Reply:
> I don't think this is a valid case

Let's say if you have a web application that just prints out chunks in a stream. In this app, you don't need `finish_reason` at all, right?

Reply:
Yes in this case we don't need `finish_reason`

But if we return the iterator data using Openai response format, then we should fill finish_reason attribute , otherwise the inconsistent format might break some applications.

If we only need to support application that prints out stream chunks, return an iterator of string is better.

Reply:
Updated: I set "finish_reason" to None in the case that we can't get finish_reason information. This is fine in open-ai sdk.